src/ppx/scene/scene_mesh.cpp

```diff
@@ -0,0 +1,142 @@
+        const uint32_t kAttributesIndex = 1;
+        mAttributeBufferView.pBuffer    = mGpuMesh->GetVertexBuffer(kAttributesIndex);
+        mAttributeBufferView.stride     = mGpuMesh->GetDerivedVertexBindings()[kAttributesIndex].GetStride();
+        mAttributeBufferView.offset     = 0;
+        mAttributeBufferView.size       = mGpuMesh->GetVertexBuffer(kAttributesIndex)->GetSize();
+    }
+}
+
+MeshData::~MeshData()
+{
+    if (mGpuMesh) {
+        auto pDevice = mGpuMesh->GetDevice();
+        pDevice->DestroyMesh(mGpuMesh);
+    }
+}
+
+// -------------------------------------------------------------------------------------------------
+// PrimitiveBatch
+// -------------------------------------------------------------------------------------------------
+PrimitiveBatch::PrimitiveBatch(
+    const scene::MaterialRef& material,
+    uint32_t                  indexOffset,
+    uint32_t                  vertexOffset,
+    uint32_t                  indexCount,
+    uint32_t                  vertexCount,
+    ppx::AABB                 boundingBox)
+    : mMaterial(material),
+      mIndexOffset(indexOffset),
+      mVertexOffset(vertexOffset),
+      mIndexCount(indexCount),
+      mVertexCount(vertexCount),
+      mBoundingBox(boundingBox)
+{
+}
+
+// -------------------------------------------------------------------------------------------------
+// Mesh
+// -------------------------------------------------------------------------------------------------
+Mesh::Mesh(
+    const scene::MeshDataRef&            meshData,
+    std::vector<scene::PrimitiveBatch>&& batches)
+    : mMeshData(meshData),
+      mBatches(std::move(batches))
+{
+    UpdateBoundingBox();
+}
+
+Mesh::Mesh(
+    std::unique_ptr<scene::ResourceManager>&& resourceManager,
+    const scene::MeshDataRef&                 meshData,
+    std::vector<scene::PrimitiveBatch>&&      batches)
+    : mResourceManager(std::move(resourceManager)),
+      mMeshData(meshData),
+      mBatches(std::move(batches))
+{
+    UpdateBoundingBox();
+}
+
+Mesh::~Mesh()
+{
+    if (mResourceManager) {
+        mResourceManager->DestroyAll();
+        mResourceManager.reset();
+    }
+}
+
+scene::VertexAttributeFlags Mesh::GetAvailableVertexAttributes() const
+{
+    return mMeshData ? mMeshData->GetAvailableVertexAttributes() : scene::VertexAttributeFlags::None();
+}
+
+void Mesh::AddBatch(const scene::PrimitiveBatch& batch)
+{
+    mBatches.push_back(batch);
+}
+
+void Mesh::UpdateBoundingBox()
+{
+    mBoundingBox.Set(float3(0));
+    for (const auto& batch : mBatches) {
+        mBoundingBox.Expand(batch.GetBoundingBox().GetMin());
+        mBoundingBox.Expand(batch.GetBoundingBox().GetMax());
+    }
+}
+
+std::vector<const scene::Material*> Mesh::GetMaterials() const
+{
+    std::vector<const scene::Material*> materials;
+    for (auto& batch : mBatches) {
+        if (IsNull(batch.GetMaterial())) {
+            // @TODO: need a better way to handle missing materials
```

Comment:
It's not necessarily an error, possibly a warning. Need to see how it plays out when we do encounter thise case. This case shuold never happen from a loader since the loader is required to set ErrorMaterial if there's missing material. But if scenes are created dynamically, it's feasible that someone could accidentally not assign a material.